core/src/local/persistent_state.rs

```diff
@@ -0,0 +1,1065 @@
+use std::borrow::Cow;
+use std::rc::Rc;
+
+use bincode;
+use itertools::Itertools;
+use rocksdb::{self, ColumnFamily, SliceTransform, WriteBatch};
+
+use ::*;
+use data::SizeOf;
+
+// Incremented on each PersistentState initialization so that IndexSeq
+// can be used to create unique identifiers for rows.
+type IndexEpoch = u64;
+
+// Monotonically increasing sequence number since last IndexEpoch used to uniquely identify a row.
+type IndexSeq = u64;
+
+// RocksDB key used for storing meta information (like indices).
+const META_KEY: &'static [u8] = b"meta";
+const META_CF: &'static str = "meta";
+
+const EXISTING_CF_ERROR: &'static str = "Invalid argument: Column family already exists";
+
+// Maximum rows per WriteBatch when building new indices for existing rows.
+const INDEX_BATCH_SIZE: usize = 512;
+
+// Store index information in RocksDB to avoid rebuilding indices on recovery.
+#[derive(Default, Serialize, Deserialize)]
+struct PersistentMeta {
+    indices: Vec<Vec<usize>>,
+    epoch: IndexEpoch,
+}
+
+#[derive(Clone)]
+struct PersistentIndex {
+    column_family: ColumnFamily,
```

Comment:
@jonhoo: had to add `impl unsafe Send` to [`ColumnFamily`](https://github.com/spacejam/rust-rocksdb/blob/701669d0763a0fb30708d36b1af2e152bda1e63d/src/lib.rs#L203-L208) in rust-rocksdb for this to work, but I think it should be fine as rust-rocksdb already stores `ColumnFamily` objects in a `BTreeMap` [internally](https://github.com/spacejam/rust-rocksdb/blob/701669d0763a0fb30708d36b1af2e152bda1e63d/src/lib.rs#L81), and the `DB` object where they're stored in is `unsafe Send` itself.

The alternative would be to just rely on rust-rocksdb's [`cf_handle`](https://github.com/spacejam/rust-rocksdb/blob/701669d0763a0fb30708d36b1af2e152bda1e63d/src/db.rs#L881-L884), which clones the `*mut ffi::rocksdb_column_family_handle_t` pointer. It takes a `&str` though, but could just keep something like a `vec!["0", 1", "2", ... to self.indices.len()]` around on `PersistentState` to avoid having to allocate through `to_string()` all the time.

Reply:
Yeah, no, that seems fine then.